### ISSUE TITLE
Fix returned tensor type for binary operators between integral tensors and floating scalars

### ISSUE THREAD

**Issue by li-roy** [pytorch/pytorch#6196]
Example below. We should probably return a floating tensor like numpy does.

```
>>> import torch
>>> a = torch.tensor([1,2,3], dtype=torch.long)
>>> a + 0.5

 1
 2
 3
[torch.LongTensor of size (3,)]
```

**Comment by goldsborough** [comment]
CC @gchanan -- any thoughts on type promotion?

**Comment by li-roy** [comment]
I was thinking maybe we could do something like #5864.

**Comment by ailzhang** [comment]
Let's use this thread to track the progress of type promotion by @colesbury. Here are some examples from other related posts. 

```
In [1]: import torch

In [2]: a = torch.LongTensor([1,2,3,4])

In [3]: a.sum() / 0.001
Floating point exception (core dumped)

## a.sum() * 0.4  also gives wrong results compared to numpy
```

**Comment by vadimkantorov** [comment]
Also multiplication of float tensor by byte tensor (mask) without explicit byte->float conversion would be nice

**Comment by zhijing-jin** [comment]
[Another Bug Report] Also Dropout works strangely with LongTensor/ Long->FloatTensor:

```
>>> a
tensor([ 100,    3,  100])
>>> a.float()
tensor([ 100.,    3.,  100.])
>>> d
Dropout(p=0.5)
>>> d(a.float())
tensor([ 200.,    0.,    0.]) # strange behavior
>>> d(a)
Floating point exception # throws error
```

**Comment by youkaichao** [comment]
If type promotion is not done, maybe we should raise a warning when type degradation happens: https://github.com/pytorch/pytorch/issues/31618

**Comment by ngimel** [comment]
This issue is fixed with the current type promotion rules.

**Comment by youkaichao** [comment]
@ngimel  hi, can you point out the version or the pull-request that resolves it? I'm using ``1.0.1`` and it remains.

**Comment by youkaichao** [comment]
Oh, I see, it is solved in ``1.3.0``. Thanks :)
